[Shipping labels] Use stored origin address for purchased labels

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-786

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This adds functionality to use the origin address stored on the backend for purchased labels.
I also added support for using different origin and destination addresses for each shipment.

Please test this PR on an order with multiple shipments and use different addresses.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. Create a test order on the web.
2. Create multiple shipments for the order.
3. Purchase a label for one shipment.
4. Purchase another label for a different shipment, using a different store address.
5. Open the order in the app.
6. Start the "Create shipping label" flow.
7. Change the selected shipment.
8. Tap on Shipment details.
9. Verify that the "Ship from" address matches the one shown on the web.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Steps above

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add selected_origin field to StoredDataDTO for origin address handling

irfano committed:
Merge stored origin and destination addresses into shipment UI model

irfano committed:
Refactor shippingAddresses to use a list and update address handling for selected shipments

irfano committed:
Refactor shippingAddresses parameter to use a list

irfano committed:
Use stored origin address when updating purchased shipment labels

irfano committed:
Ensure default destination address is set only when empty and not verified

irfano committed:
Update WooShippingLabelCreationScreen to use list shippingAddresses

irfano committed:
Fix tests selected_origin update

irfano committed:
Use different destination address for each shipments

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/262">22.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

irfano committed:
Revert "Use different destination address for each shipments"

This reverts commit 9f350c37470f78f9219bea925bcd621d342e4071.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/GetShipments.kt:
```diff
@@ -91,17 +96,26 @@ class GetShipments @Inject constructor(
     }
 
-    private fun mergeDestinationAddresses(
+    private fun mergeAddresses(
```
We were already adding the stored destination address to the shipments here. Now, we’re also adding the origin address in the same way.

JorgeMucientes replied:
Mino np, this function could be slightly simplified making it a bit more readable? Feel free to keep things as they are if you prefer the existing code. 

```kotlin
    private fun mergeAddresses(
        shipments: List<ShipmentUIModel>,
        originAddresses: Map<String, OriginAddressDTO>?,
        destinationAddresses: Map<String, DestinationAddressDTO>?
    ): List<ShipmentUIModel> = shipments.map { shipment ->
        val remoteId = shipment.remoteId ?: return@map shipment
        val key = getStoredDataKey(remoteId)

        val updatedLabel = shipment.label
            ?.copy(
                originAddress = originAddresses?.get(key)?.let { mapper(it) } 
                    ?: shipment.label.originAddress,
                destinationAddress = destinationAddresses?.get(key)?.let { mapper(it) }
                    ?: shipment.label.destinationAddress
            )
        shipment.copy(label = updatedLabel)
    }
```

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -127,7 +128,7 @@ class WooShippingLabelCreationViewModel @Inject constructor(
     private val emptyOrder = Order.getEmptyOrder(Date(), Date())
     private val order = MutableStateFlow(emptyOrder)
     private val destinationAddress = MutableStateFlow(DestinationShippingAddress.EMPTY)
-    private val shippingAddresses = MutableStateFlow<WooShippingAddresses?>(WooShippingAddresses.EMPTY)
+    private val shippingAddresses = MutableStateFlow<List<WooShippingAddresses>>(emptyList())
```
Since we have a separate shippingAddresses for each shipment, I converted `shippingAddresses` to a list, similar to how we manage other shipment-related data such as `selectedPackagesFlow`, `customsFormDataFlow`, `packageWeightsFlow`...

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -255,9 +254,11 @@ class WooShippingLabelCreationViewModel @Inject constructor(
                 // If result has a label model update the label with it. Otherwise, just update the status.
                 val newLabel = result.shippingLabelModel ?: shipment.label.copy(status = result.status)
-                updateShipment(shipmentId, shipment.copy(label = newLabel))
+                updateShipment(shipmentId, shipment.copy(label = newLabel.copy(originAddress = originAddress)))
```
`observeShippingLabelStatus ` observes the status of the purchase, but the origin address is not included in the label it returns. To make it easier to manage in the UI, we added the origin address to the label model.
So when using the `label` from the observeShippingLabelStatus endpoint, I'm adding the current origin address (used during the label purchase) to the label model.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -271,14 +272,17 @@ class WooShippingLabelCreationViewModel @Inject constructor(
-                if (addressValidationHelper.isMissingDestinationAddress(order.shippingAddress).not()) {
+                if (addressValidationHelper.isMissingDestinationAddress(order.shippingAddress).not() &&
+                    !destinationAddress.value.isVerified
+                ) {
```
I fixed a bug that was introduced in #14272. In that PR, I added `uiState.map { it.selectedIndex }.distinctUntilChanged()` to the `combine` function, which causes `getDestinationAddress()` to be triggered each time the selected shipment changes.
However, we shouldn’t reset the destinationAddress to the default on each change, it should only be set once when the view model is initialized.
We should also call `verifyDestinationAddress` only if the address hasn’t already been verified. This bug was causing unnecessary “Verified destination” notices.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -458,17 +465,46 @@ class WooShippingLabelCreationViewModel @Inject constructor(
+                    }
+
+                WooShippingAddresses(
+                    shipFrom = shipFrom ?: OriginShippingAddress.EMPTY,
```
Ideally, `shipFrom` shouldn't be `null`. But if it's due to some backend issue, we're showing the empty address. This behavior is also the same as how web currently behaves.

JorgeMucientes replied:
Do you have the link to the discussion around this backend issue? I'd like to understand why and when this happens 🙏🏼

irfano replied:
I’m not sure if this was discussed, but you can test order #2006 on our testing site to see the empty origin address on the map. It might be caused by an incomplete feature from the mobile sending incorrect data to the backend, so we’ll need to test this during the Milestone 4 testing phase.
That said, for any potential `null` `shipFrom` case, it looks good to me to show an empty address.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -458,17 +465,46 @@ class WooShippingLabelCreationViewModel @Inject constructor(
+
+                WooShippingAddresses(
+                    shipFrom = shipFrom ?: OriginShippingAddress.EMPTY,
+                    originAddresses = originAddresses.orEmpty(),
```
Before this PR, if `originAddresses` was empty, we were displaying an error screen. But I thought that’s not necessary, users should still be able to see purchased labels even if the site doesn’t have a store address.
Also, I couldn’t test this, as it’s likely not possible to enter this flow when originAddresses is empty.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -553,24 +589,22 @@ class WooShippingLabelCreationViewModel @Inject constructor(
                 return@combine WooShippingViewState.Error
             }
 
+            shipmentItems.value = shipments.map { it.items }
+            adjustFlowSizesToShipmentCount(shipments.size)
```
I just moved these lines above the `destinationStatus` update.

irfano commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModelTest.kt:
```diff
@@ -379,18 +379,6 @@ class WooShippingLabelCreationViewModelTest : BaseUnitTest() {
         assertNotNull(exit)
     }
 
-    @Test
-    fun `when there are no origin addresses, then show an error`() = testBlocking {
-        whenever(observeOriginAddresses()) doReturn flowOf(emptyList())
-
-        createViewModel()
-
-        advanceUntilIdle()
-
-        val currentViewState = sut.viewState.value
-        assert(currentViewState is WooShippingViewState.Error)
-    }
-
```
This case is no longer valid, we no longer show an error and if there is no origin address. Now ,users can see purchase labels instead of an error screen (I mentioned [here](https://github.com/woocommerce/woocommerce-android/pull/14300/files#r2197334148)).

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingNetworkingMapper.kt:
```diff
@@ -91,6 +92,26 @@ class WooShippingNetworkingMapper @Inject constructor(
         )
     }
 
+    operator fun invoke(originAddressDTO: OriginAddressDTO): Address {
+        val name = originAddressDTO.name?.split(" ") ?: listOf("", "")
+        return Address(
+            company = originAddressDTO.company.orEmpty(),
+            firstName = name.getOrElse(0) { "" },
+            lastName = name.getOrElse(1) { "" },
+            phone = originAddressDTO.phone.orEmpty(),
+            country = Location(
+                name = originAddressDTO.country.orEmpty(),
+                code = originAddressDTO.country.orEmpty()
+            ),
+            state = AmbiguousLocation.Raw(originAddressDTO.state.orEmpty()),
+            address1 = originAddressDTO.address.orEmpty(),
+            address2 = originAddressDTO.address2.orEmpty(),
+            city = originAddressDTO.city.orEmpty(),
+            postcode = originAddressDTO.postcode.orEmpty(),
+            email = ""
+        )
+    }
+
```
Minor np. 
Isn't this duplicated with `operator fun invoke(destinationAddressDTO: DestinationAddressDTO): Address {` ?
It seems to me the only difference is the input parameter for this mapping functions and when checking `DestinationAddressDTO` and `OriginAddressDTO` they look the same (except for @SerializedName address_2 and email which is an optional field anyway). Wdyt about merging those models (we can address this in a different PR) and use the same DTO and mapping functions to remove duplication?

irfano replied:
Good observation. We have an open issue to refactor the models: WOOMOB-689

JorgeMucientes reviewed (approved):
Nicely done @irfano I just left a few suggestion around the code but nothing blocking. Changes look good and work as expected

irfano committed:
Add JsonAdapter for selected_origin and selected_destination in StoredDataDTO

irfano committed:
Simplify ShipmentMapDeserializer

irfano committed:
Simplify mergeAddresses() function

harrytmthy commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingNetworkingMapper.kt:
```diff
@@ -91,6 +92,26 @@ class WooShippingNetworkingMapper @Inject constructor(
     }
 
+    operator fun invoke(originAddressDTO: OriginAddressDTO): Address {
+        val name = originAddressDTO.name?.split(" ") ?: listOf("", "")
```
Hi @irfano, splitting `name` using `.split(" ")` will incorrectly assign first/last names if the name has more than two words. For example, `"John Michael Doe"` would result in:
- **First name:** John
- **Last name:** Michael

instead of the expected:
- **First name:** John Michael
- **Last name:** Doe

A simple approach is to use `indexOfLast { it == ' ' }` with substring:

```kotlin
val name = originAddressDTO.name.orEmpty()
val lastSpaceIndex = name.indexOfLast { it == ' ' }
val (firstName, lastName) = if (lastSpaceIndex == -1) {
    name to ""
} else {
    name.substring(0, lastSpaceIndex) to name.substring(lastSpaceIndex + 1)
}
```

Hope it helps 🙂

irfano replied:
Good point! Thanks for the suggestion. I applied it to the change introduced by this PR and to other functions that use the same method. 🙏🏻 (c62981a7ba21df834158a319abda5acbeffed288)

We actually can’t know if the last name is the last word or the last two words, and I’m not sure why it’s necessary to parse first and last names. But still, this suggestion is better than the original one.

irfano committed:
Improve parsing of full names in WooShippingNetworkingMapper

This commit introduces a new `parseFullName` private method to handle the splitting of full names into first and last names.

The new method improves upon the previous logic by:
- Correctly handling names with multiple spaces.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-786-use-stored-origin-address-for-purchased-labels

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingNetworkingMapper.kt

irfano committed:
Fix detekt errors
